Bug: processor chain crashes after filter

### Bug description

Any processor that executes after `filter` will crash with:

```
ERR pipeline worker stopped error="task outbox-events:source:enrich_metadata: processor didn't return any records"
```

As explained by @hariso, this is because a filter record is of a special type.

### Steps to reproduce

Example pipeline:

```
version: 2.2
pipelines:
  - id: outbox-events
    status: running
    connectors:
      - id: source
        type: source
        plugin: builtin:postgres
        settings:
          url: ${PG_SOURCE_DSN}
          tables: outbox_events
          cdcMode: logrepl
          logrepl.autoCleanup: 'false'
          logrepl.publicationName: cnd_outbox
          logrepl.slotName: cnd_outbox_events
          logrepl.withAvroSchema: 'false'
          snapshot.fetchSize: '1000'
          snapshotMode: initial
        processors:
          - id: insert_only
            plugin: 'filter'
            condition: '{{ ne .Operation 1 }}'
          - id: enrich_metadata
            plugin: 'custom.javascript'
            settings:
              script: |
                function process(rec) {
                  return rec;
                }
     - id: output-log
       type: destination
       plugin: builtin:log
       settings:
         sdk.schema.extract.key.enabled: 'false'
         sdk.schema.extract.payload.enabled: 'false'
```

### Version

v0.13.4

Another issue related to this, but much more serious.

```
panic: runtime error: index out of range [-1]

goroutine 2139 [running]:
github.com/conduitio/conduit-connector-sdk.(*destinationPluginAdapter).Run(0xc0018da690, {0x2ef6af8?, 0xc0015e5980?}, {0x2ee4410, 0xc000d24268})
 github.com/conduitio/conduit-connector-sdk@v0.13.3/destination.go:202 +0x294
github.com/conduitio/conduit/pkg/plugin/connector/builtin.(*destinationPluginAdapter).Run.func1()
 github.com/conduitio/conduit@v0.13.4/pkg/plugin/connector/builtin/destination.go:84 +0xf0
created by github.com/conduitio/conduit/pkg/plugin/connector/builtin.(*destinationPluginAdapter).Run in goroutine 1
 github.com/conduitio/conduit@v0.13.4/pkg/plugin/connector/builtin/destination.go:83 +0x29b
```

This happens when destination is not receiving any records because they've all been filtered out.

--- a/pkg/lifecycle-poc/funnel/batch.go
+++ b/pkg/lifecycle-poc/funnel/batch.go
@@ -58,6 +58,15 @@ func NewBatch(records []opencdc.Record) *Batch {
 	}
 }
 
+// Ack marks the record at index i as acked. If a second index is provided, all
+// records between i (included) and j (excluded) are marked as acked. If multiple
+// indices are provided, the method panics.
+// Records are marked as acked by default, so this method is only useful when
+// reprocessing records marked to be retried.
+func (b *Batch) Ack(i int, j ...int) {
+	b.setFlagNoErr(RecordFlagAck, i, j...)
+}
+
 // Nack marks the record at index i as nacked. If multiple errors are provided,
 // they are assigned to the records starting at index i.
 func (b *Batch) Nack(i int, errs ...error) {
@@ -66,8 +75,8 @@ func (b *Batch) Nack(i int, errs ...error) {
 }
 
 // Retry marks the record at index i to be retried. If a second index is
-// provided, all records between i (included) and j (excluded) are marked as
-// acked. If multiple indices are provided, the method panics.
+// provided, all records between i (included) and j (excluded) are marked to be
+// retried. If multiple indices are provided, the method panics.
 func (b *Batch) Retry(i int, j ...int) {
 	b.setFlagNoErr(RecordFlagRetry, i, j...)
 	b.tainted = true

--- a/pkg/lifecycle-poc/funnel/destination.go
+++ b/pkg/lifecycle-poc/funnel/destination.go
@@ -84,6 +84,9 @@ func (t *DestinationTask) Close(ctx context.Context) error {
 
 func (t *DestinationTask) Do(ctx context.Context, batch *Batch) error {
 	records := batch.ActiveRecords()
+
+	// Store the positions of the records in the batch to be used for
+	// validation of acks.
 	positions := make([]opencdc.Position, len(records))
 	for i, rec := range records {
 		positions[i] = rec.Position
@@ -95,31 +98,51 @@ func (t *DestinationTask) Do(ctx context.Context, batch *Batch) error {
 		return cerrors.Errorf("failed to write %d records to destination: %w", len(positions), err)
 	}
 
-	acks := make([]connector.DestinationAck, 0, len(positions))
+	activeIndices := batch.ActiveRecordIndices()
+	var ackCount = 0
 	for range len(positions) {
-		acksResp, err := t.destination.Ack(ctx)
+		acks, err := t.destination.Ack(ctx)
 		if err != nil {
 			return cerrors.Errorf("failed to receive acks for %d records from destination: %w", len(positions), err)
 		}
-		t.metrics.Observe(records[len(acks):len(acks)+len(acksResp)], start)
-		acks = append(acks, acksResp...)
-		if len(acks) >= len(positions) {
-			break
-		}
-	}
-	if len(acks) != len(positions) {
-		return cerrors.Errorf("received %d acks, but expected %d", len(acks), len(positions))
-	}
+		t.metrics.Observe(records[ackCount:ackCount+len(acks)], start)
 
-	for i, ack := range acks {
-		if !bytes.Equal(positions[i], ack.Position) {
-			// TODO is this a fatal error? Looks like a bug in the connector
-			return cerrors.Errorf("received unexpected ack, expected position %q but got %q", positions[i], ack.Position)
+		if err := t.validateAcks(acks, positions[ackCount:]); err != nil {
+			return cerrors.Errorf("failed to validate acks: %w", err)
 		}
-		if ack.Error != nil {
-			batch.Nack(i, ack.Error) // TODO: adjust i according to filtered records
+		t.markBatchRecords(batch, activeIndices, ackCount, acks)
+
+		ackCount += len(acks)
+		if ackCount >= len(positions) {
+			break
 		}
 	}
 
 	return nil
+}
+
+func (t *DestinationTask) validateAcks(acks []connector.DestinationAck, positions []opencdc.Position) error {
+	if len(acks) > len(positions) {
+		return cerrors.Errorf("received %d acks, but expected at most %d", len(acks), len(positions))
+	}
+
+	for i, ack := range acks {
+		if !bytes.Equal(positions[i], ack.Position) {
+			return cerrors.Errorf("received unexpected ack, expected position %q but got %q", positions[i], ack.Position)
+		}
+	}
+
+	return nil
+}
+
+func (t *DestinationTask) markBatchRecords(b *Batch, activeIndices []int, from int, acks []connector.DestinationAck) {
+	for i, ack := range acks {
+		if ack.Error != nil {
+			idx := from + i
+			if activeIndices != nil {
+				idx = activeIndices[idx]
+			}
+			b.Nack(from+i, ack.Error)
+		}
+	}
 }

--- a/pkg/lifecycle-poc/funnel/destination.go
+++ b/pkg/lifecycle-poc/funnel/destination.go
@@ -99,17 +99,17 @@ func (t *DestinationTask) Do(ctx context.Context, batch *Batch) error {
 	}
 
 	activeIndices := batch.ActiveRecordIndices()
-	var ackCount = 0
+	ackCount := 0
 	for range len(positions) {
 		acks, err := t.destination.Ack(ctx)
 		if err != nil {
 			return cerrors.Errorf("failed to receive acks for %d records from destination: %w", len(positions), err)
 		}
-		t.metrics.Observe(records[ackCount:ackCount+len(acks)], start)
 
 		if err := t.validateAcks(acks, positions[ackCount:]); err != nil {
 			return cerrors.Errorf("failed to validate acks: %w", err)
 		}
+		t.metrics.Observe(records[ackCount:ackCount+len(acks)], start)
 		t.markBatchRecords(batch, activeIndices, ackCount, acks)
 
 		ackCount += len(acks)
@@ -142,7 +142,7 @@ func (t *DestinationTask) markBatchRecords(b *Batch, activeIndices []int, from i
 			if activeIndices != nil {
 				idx = activeIndices[idx]
 			}
-			b.Nack(from+i, ack.Error)
+			b.Nack(idx, ack.Error)
 		}
 	}
 }

--- a/pkg/lifecycle-poc/funnel/destination_test.go
+++ b/pkg/lifecycle-poc/funnel/destination_test.go
@@ -1,0 +1,58 @@
+// Copyright © 2025 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package funnel
+
+import (
+	"context"
+	"testing"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/lifecycle-poc/funnel/mock"
+	"github.com/matryer/is"
+	"go.uber.org/mock/gomock"
+)
+
+func TestDestinationTask_Do_Success(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+	destinationMock.EXPECT().Write(ctx, records).Return(nil)
+
+	acks := make([]connector.DestinationAck, len(records))
+	for i, record := range records {
+		acks[i] = connector.DestinationAck{
+			Position: record.Position,
+			Error:    nil,
+		}
+	}
+
+	destinationMock.EXPECT().Ack(ctx).Return(acks, nil)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+	err := task.Do(ctx, batch)
+	is.NoErr(err)
+
+	is.Equal(batch.ActiveRecords(), records)
+	for _, status := range batch.recordStatuses {
+		is.Equal(status, RecordStatus{Flag: RecordFlagAck})
+	}
+}

--- a/pkg/lifecycle-poc/funnel/destination_test.go
+++ b/pkg/lifecycle-poc/funnel/destination_test.go
@@ -17,15 +17,19 @@ package funnel
 import (
 	"context"
 	"testing"
+	"time"
 
+	"github.com/conduitio/conduit-commons/csync"
+	"github.com/conduitio/conduit-commons/opencdc"
 	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
 	"github.com/conduitio/conduit/pkg/foundation/log"
 	"github.com/conduitio/conduit/pkg/lifecycle-poc/funnel/mock"
 	"github.com/matryer/is"
 	"go.uber.org/mock/gomock"
 )
 
-func TestDestinationTask_Do_Success(t *testing.T) {
+func TestDestinationTask_Do_Simple(t *testing.T) {
 	is := is.New(t)
 	ctx := context.Background()
 	logger := log.Test(t)
@@ -55,4 +59,214 @@ func TestDestinationTask_Do_Success(t *testing.T) {
 	for _, status := range batch.recordStatuses {
 		is.Equal(status, RecordStatus{Flag: RecordFlagAck})
 	}
+}
+
+func TestDestinationTask_Do_WriteFailure(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+
+	wantErr := cerrors.New("error")
+	destinationMock.EXPECT().Write(ctx, records).Return(wantErr)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+	err := task.Do(ctx, batch)
+	is.True(cerrors.Is(err, wantErr))
+}
+
+func TestDestinationTask_Do_FilteredAndNacks(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+
+	// Records with index 0, 2, 3 and 9 are filtered out
+	batch.Filter(0)    // 0
+	batch.Filter(2, 4) // 2, 3
+	batch.Filter(9)    // 9
+
+	activeRecords := batch.ActiveRecords()
+	is.Equal(activeRecords, []opencdc.Record{records[1], records[4], records[5], records[6], records[7], records[8]})
+
+	destinationMock.EXPECT().Write(ctx, activeRecords).Return(nil)
+
+	acks := make([]connector.DestinationAck, len(activeRecords))
+	wantErr := cerrors.New("error")
+	for i, record := range activeRecords {
+		ack := connector.DestinationAck{
+			Position: record.Position,
+			Error:    nil,
+		}
+		if i == 0 || i == 2 || i == 4 {
+			// records 1, 5 and 7 are nacked
+			ack.Error = wantErr
+		}
+		acks[i] = ack
+	}
+
+	destinationMock.EXPECT().Ack(ctx).Return(acks, nil)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+	err := task.Do(ctx, batch)
+	is.NoErr(err)
+
+	is.Equal(batch.ActiveRecords(), activeRecords)
+	for i, status := range batch.recordStatuses {
+		switch i {
+		case 0, 2, 3, 9:
+			is.Equal(status, RecordStatus{Flag: RecordFlagFilter})
+		case 1, 5, 7:
+			is.Equal(status, RecordStatus{Flag: RecordFlagNack, Error: wantErr})
+		default:
+			is.Equal(status, RecordStatus{Flag: RecordFlagAck})
+		}
+	}
+}
+
+func TestDestinationTask_Do_MultipleAcks(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+	destinationMock.EXPECT().Write(ctx, records).Return(nil)
+
+	acks := make([]connector.DestinationAck, len(records))
+	for i, record := range records {
+		acks[i] = connector.DestinationAck{
+			Position: record.Position,
+			Error:    nil,
+		}
+	}
+
+	call := destinationMock.EXPECT().Ack(ctx).Return(acks[0:3], nil).Call
+	call = destinationMock.EXPECT().Ack(ctx).Return(acks[3:6], nil).After(call)
+	call = destinationMock.EXPECT().Ack(ctx).Return(acks[6:7], nil).After(call)
+	destinationMock.EXPECT().Ack(ctx).Return(acks[7:10], nil).After(call)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+	err := task.Do(ctx, batch)
+	is.NoErr(err)
+
+	is.Equal(batch.ActiveRecords(), records)
+	for _, status := range batch.recordStatuses {
+		is.Equal(status, RecordStatus{Flag: RecordFlagAck})
+	}
+}
+
+func TestDestinationTask_Do_UnexpectedPosition(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+	destinationMock.EXPECT().Write(ctx, records).Return(nil)
+
+	acks := make([]connector.DestinationAck, len(records))
+	for i, record := range records {
+		acks[i] = connector.DestinationAck{
+			Position: record.Position,
+			Error:    nil,
+		}
+	}
+
+	call := destinationMock.EXPECT().Ack(ctx).Return(acks[0:3], nil).Call
+	// Note that we're skipping record 3 here
+	destinationMock.EXPECT().Ack(ctx).Return(acks[4:8], nil).After(call)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+	err := task.Do(ctx, batch)
+	is.Equal(err.Error(), `failed to validate acks: received unexpected ack, expected position "3" but got "4"`)
+}
+
+func TestDestinationTask_Do_TooManyAcks(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+	destinationMock.EXPECT().Write(ctx, records).Return(nil)
+
+	// Note that we're creating more acks than records
+	acks := make([]connector.DestinationAck, len(records)+1)
+
+	destinationMock.EXPECT().Ack(ctx).Return(acks, nil)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+	err := task.Do(ctx, batch)
+	is.Equal(err.Error(), `failed to validate acks: received 11 acks, but expected at most 10`)
+}
+
+func TestDestinationTask_Do_FailedToReceiveAcks(t *testing.T) {
+	is := is.New(t)
+	ctx := context.Background()
+	logger := log.Test(t)
+
+	ctrl := gomock.NewController(t)
+	destinationMock := mock.NewDestination(ctrl)
+
+	records := randomRecords(10)
+	batch := NewBatch(records)
+	destinationMock.EXPECT().Write(gomock.Any(), records).Return(nil)
+
+	acks := make([]connector.DestinationAck, len(records))
+	for i, record := range records {
+		acks[i] = connector.DestinationAck{
+			Position: record.Position,
+			Error:    nil,
+		}
+	}
+
+	// Note that we're returning only the first 9 acks, the second call blocks forever
+	call := destinationMock.EXPECT().Ack(gomock.Any()).Return(acks[:9], nil).Call
+	destinationMock.EXPECT().Ack(gomock.Any()).DoAndReturn(func(ctx context.Context) ([]connector.DestinationAck, error) {
+		<-ctx.Done()
+		return nil, ctx.Err()
+	}).After(call)
+
+	task := NewDestinationTask("test", destinationMock, logger, NoOpConnectorMetrics{})
+
+	// We need to run Do in a goroutine, as we expect it to block forever
+	var wg csync.WaitGroup
+	wg.Add(1)
+	var doErr error
+
+	doCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go func() {
+		defer wg.Done()
+		doErr = task.Do(doCtx, batch)
+	}()
+
+	err := wg.WaitTimeout(ctx, 100*time.Millisecond)
+	is.Equal(err, context.DeadlineExceeded)
+
+	cancel()
+	err = wg.WaitTimeout(ctx, 100*time.Millisecond)
+	is.NoErr(err)
+	is.Equal(doErr.Error(), `failed to receive acks for 10 records from destination: context canceled`)
 }

--- a/pkg/lifecycle-poc/funnel/funnel_test.go
+++ b/pkg/lifecycle-poc/funnel/funnel_test.go
@@ -63,9 +63,11 @@ func Example_simpleStream() {
 		&NoOpConnectorMetrics{},
 	)
 
+	destTaskNode := &TaskNode{Task: destTask}
+	srcTaskNode := &TaskNode{Task: srcTask, Next: []*TaskNode{destTaskNode}}
+
 	w, err := NewWorker(
-		[]Task{srcTask, destTask},
-		[][]int{{1}, {}},
+		srcTaskNode,
 		dlq,
 		logger,
 		noop.Timer{},
@@ -167,9 +169,11 @@ func BenchmarkStreamNew(b *testing.B) {
 			&NoOpConnectorMetrics{},
 		)
 
+		destTaskNode := &TaskNode{Task: destTask}
+		srcTaskNode := &TaskNode{Task: srcTask, Next: []*TaskNode{destTaskNode}}
+
 		w, err := NewWorker(
-			[]Task{srcTask, destTask},
-			[][]int{{1}, {}},
+			srcTaskNode,
 			dlq,
 			logger,
 			noop.Timer{},

--- a/pkg/lifecycle-poc/funnel/worker.go
+++ b/pkg/lifecycle-poc/funnel/worker.go
@@ -16,6 +16,7 @@ package funnel
 
 import (
 	"context"
+	"iter"
 	"sync/atomic"
 	"time"
 
@@ -73,19 +74,9 @@ type Task interface {
 // The rest of the records are processed as a sub-batch, and the same rules
 // apply to them.
 type Worker struct {
-	Source Source
-	Tasks  []Task
-	// Order defines the next task to be executed. Multiple indices are used to
-	// show parallel execution of tasks.
-	//
-	// Example:
-	// [[1], [2], [3,5], [4], [], []]
-	//
-	//            /-> 3 -> 4
-	// 0 -> 1 -> 2
-	//            \-> 5
-	Order Order
-	DLQ   *DLQ
+	Source    Source
+	FirstTask *TaskNode
+	DLQ       *DLQ
 
 	lastReadAt time.Time
 	timer      metrics.Timer
@@ -100,70 +91,45 @@ type Worker struct {
 }
 
 func NewWorker(
-	tasks []Task,
-	order Order,
+	firstTask *TaskNode,
 	dlq *DLQ,
 	logger log.CtxLogger,
 	timer metrics.Timer,
 ) (*Worker, error) {
-	err := validateTaskOrder(tasks, order)
+	firstTask.first = true // mark the first task as the first task in the pipeline
+	err := validateTasks(firstTask)
 	if err != nil {
 		return nil, cerrors.Errorf("invalid task order: %w", err)
 	}
 
-	st, ok := tasks[0].(interface{ GetSource() Source })
+	st, ok := firstTask.Task.(interface{ GetSource() Source })
 	if !ok {
-		return nil, cerrors.Errorf("first task must be a source task, got %T", tasks[0])
+		return nil, cerrors.Errorf("first task must be a source task, got %T", firstTask.Task)
 	}
 
 	return &Worker{
-		Source: st.GetSource(),
-		Tasks:  tasks,
-		Order:  order,
-		DLQ:    dlq,
-		logger: logger.WithComponent("funnel.Worker"),
-		timer:  timer,
+		Source:    st.GetSource(),
+		FirstTask: firstTask,
+		DLQ:       dlq,
+		logger:    logger.WithComponent("funnel.Worker"),
+		timer:     timer,
 
 		processingLock: make(chan struct{}, 1),
 	}, nil
 }
 
-func validateTaskOrder(tasks []Task, order Order) error {
+func validateTasks(task *TaskNode) error {
 	// Traverse the tasks according to the order and validate that each task
 	// is included exactly once.
-	if len(order) != len(tasks) {
-		return cerrors.Errorf("order length (%d) does not match tasks length (%d)", len(order), len(tasks))
-	}
-	seenCount := make([]int, len(tasks))
-	var traverse func(i int) error
-	traverse = func(i int) error {
-		if i < 0 || i >= len(tasks) {
-			return cerrors.Errorf("invalid index (%d), expected a number between 0 and %d", i, len(tasks)-1)
+	seen := make(map[string]bool)
+
+	for t := range task.Tasks() {
+		if seen[t.ID()] {
+			return cerrors.Errorf("task %s included multiple times in order", task.Task.ID())
 		}
-		seenCount[i]++
-		if seenCount[i] > 1 {
-			return cerrors.Errorf("task %d included multiple times in order", i)
-		}
-		for _, nextIdx := range order[i] {
-			if nextIdx == i {
-				return cerrors.Errorf("task %d cannot call itself as next task", i)
-			}
-			err := traverse(nextIdx)
-			if err != nil {
-				return err
-			}
-		}
-		return nil
+		seen[t.ID()] = true
 	}
-	err := traverse(0)
-	if err != nil {
-		return err
-	}
-	for i, count := range seenCount {
-		if count == 0 {
-			return cerrors.Errorf("task %d not included in order", i)
-		}
-	}
+
 	return nil
 }
 
@@ -180,7 +146,7 @@ func (w *Worker) Open(ctx context.Context) (err error) {
 		})
 	}()
 
-	for _, task := range w.Tasks {
+	for task := range w.FirstTask.Tasks() {
 		err = task.Open(ctx)
 		if err != nil {
 			return cerrors.Errorf("task %s failed to open: %w", task.ID(), err)
@@ -239,7 +205,7 @@ func (w *Worker) acquireProcessingLock(ctx context.Context) (release func(), err
 func (w *Worker) Close(ctx context.Context) error {
 	var errs []error
 
-	for _, task := range w.Tasks {
+	for task := range w.FirstTask.Tasks() {
 		err := task.Close(ctx)
 		if err != nil {
 			errs = append(errs, cerrors.Errorf("task %s failed to close: %w", task.ID(), err))
@@ -259,7 +225,7 @@ func (w *Worker) Close(ctx context.Context) error {
 func (w *Worker) Do(ctx context.Context) error {
 	for !w.stop.Load() {
 		w.logger.Trace(ctx).Msg("starting next batch")
-		if err := w.doTask(ctx, 0, &Batch{}, w); err != nil {
+		if err := w.doTask(ctx, w.FirstTask, &Batch{}, w); err != nil {
 			return err
 		}
 		w.logger.Trace(ctx).Msg("batch done")
@@ -268,8 +234,13 @@ func (w *Worker) Do(ctx context.Context) error {
 }
 
 //nolint:gocyclo // TODO: refactor
-func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker ackNacker) error {
-	t := w.Tasks[currentIndex]
+func (w *Worker) doTask(
+	ctx context.Context,
+	taskNode *TaskNode,
+	b *Batch,
+	acker ackNacker,
+) error {
+	t := taskNode.Task
 
 	w.logger.Trace(ctx).
 		Str("task_id", t.ID()).
@@ -290,14 +261,14 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 		// ErrPluginNotRunning can be returned if the plugin is stopped before
 		// trying to read the next batch.
 		// Both are considered as graceful stop, just return the context error, if any.
-		if currentIndex == 0 && (cerrors.Is(err, context.Canceled) ||
+		if w.isFirstTask(t) && (cerrors.Is(err, context.Canceled) ||
 			(cerrors.Is(err, plugin.ErrPluginNotRunning) && w.stop.Load())) {
 			return ctx.Err()
 		}
 		return cerrors.Errorf("task %s: %w", t.ID(), err)
 	}
 
-	if currentIndex == 0 {
+	if w.isFirstTask(t) {
 		// The first task has some specifics:
 		// - Store last time we read a batch from the source for metrics.
 		// - It locks the stop lock, so that no stop signal can be received while
@@ -330,13 +301,13 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 			Msg("task returned clean batch")
 
 		// Shortcut.
-		if !w.hasNextTask(currentIndex) || !b.HasActiveRecords() {
+		if !taskNode.HasNext() || !b.HasActiveRecords() {
 			// Either this is the last task (the batch has made it end-to-end),
 			// or the batch has only filtered records. Let's ack!
 			return acker.Ack(ctx, b)
 		}
 		// There is at least one task after this one, let's continue.
-		return w.doNextTask(ctx, currentIndex, b, acker)
+		return w.doNextTask(ctx, taskNode, b, acker)
 	}
 
 	w.logger.Trace(ctx).
@@ -361,7 +332,7 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 
 		switch subBatch.recordStatuses[0].Flag {
 		case RecordFlagAck, RecordFlagFilter:
-			if !w.hasNextTask(currentIndex) || !subBatch.HasActiveRecords() {
+			if !taskNode.HasNext() || !subBatch.HasActiveRecords() {
 				// Either this is the last task (the batch has made it end-to-end),
 				// or the batch has only filtered records. Let's ack!
 				// We need to ack all the records in the batch, not only active
@@ -373,7 +344,7 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 				break // break switch
 			}
 			// There is at least one task after this one, let's continue.
-			err := w.doNextTask(ctx, currentIndex, subBatch, acker)
+			err := w.doNextTask(ctx, taskNode, subBatch, acker)
 			if err != nil {
 				return err
 			}
@@ -383,7 +354,10 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 				return err
 			}
 		case RecordFlagRetry:
-			err := w.doTask(ctx, currentIndex, subBatch, acker)
+			// Retry the sub-batch by passing it to the same task. We need to
+			// mark the records as acked, as that's the default record status.
+			subBatch.Ack(0, len(subBatch.records))
+			err := w.doTask(ctx, taskNode, subBatch, acker)
 			if err != nil {
 				return err
 			}
@@ -429,19 +403,18 @@ OUTER:
 	return b.sub(firstIndex, lastIndex)
 }
 
-func (w *Worker) hasNextTask(currentIndex int) bool {
-	return len(w.Order[currentIndex]) > 0
+func (w *Worker) isFirstTask(t Task) bool {
+	return w.FirstTask.Task.ID() == t.ID()
 }
 
-func (w *Worker) doNextTask(ctx context.Context, currentIndex int, b *Batch, acker ackNacker) error {
-	nextIndices := w.Order[currentIndex]
-	switch len(nextIndices) {
+func (w *Worker) doNextTask(ctx context.Context, taskNode *TaskNode, b *Batch, acker ackNacker) error {
+	switch len(taskNode.Next) {
 	case 0:
 		// no next task, we're done
 		return nil
 	case 1:
 		// single next task, let's pass the batch to it
-		return w.doTask(ctx, nextIndices[0], b, acker)
+		return w.doTask(ctx, taskNode.Next[0], b, acker)
 	default:
 		// TODO(multi-connector): remove error
 		return cerrors.Errorf("multiple next tasks not supported yet")
@@ -449,12 +422,12 @@ func (w *Worker) doNextTask(ctx context.Context, currentIndex int, b *Batch, ack
 		// multiple next tasks, let's clone the batch and pass it to them
 		// concurrently
 		//nolint:govet // TODO implement multi ack nacker
-		multiAcker := newMultiAckNacker(acker, len(nextIndices))
+		multiAcker := newMultiAckNacker(acker, len(taskNode.Next))
 		p := pool.New().WithErrors() // TODO WithContext?
-		for _, i := range nextIndices {
+		for _, nextTask := range taskNode.Next {
 			b := b.clone()
 			p.Go(func() error {
-				return w.doTask(ctx, i, b, multiAcker)
+				return w.doTask(ctx, nextTask, b, multiAcker)
 			})
 		}
 		err := p.Wait()
@@ -509,43 +482,85 @@ func (w *Worker) updateTimer(records []opencdc.Record) {
 	}
 }
 
-// Order represents the order of tasks in a pipeline. Each index in the slice
-// represents a task, and the value at that index is a slice of indices of the
-// next tasks to be executed. If the slice is empty, the task is the last one in
-// the pipeline.
-type Order [][]int
+// TaskNode represents a task in the pipeline. It contains the task itself and
+// the next tasks to be executed after it.
+type TaskNode struct {
+	Task Task
+	Next []*TaskNode
 
-// AppendSingle appends a single element to the current order.
-func (o Order) AppendSingle(next []int) Order {
-	if len(o) == 0 {
-		return Order{next}
-	}
-	o[len(o)-1] = append(o[len(o)-1], len(o))
-	return append(o, next)
+	first bool
 }
 
-// AppendOrder appends the next order to the current order. The next order indices
-// are adjusted to match the new order length.
-func (o Order) AppendOrder(next Order) Order {
-	if len(o) == 0 {
-		return next
-	} else if len(next) == 0 {
-		return o
-	}
-
-	next.Increase(len(o))
-	o[len(o)-1] = append(o[len(o)-1], len(o))
-	return append(o, next...)
+// IsFirst returns true if this task is the first task in the pipeline.
+func (t *TaskNode) IsFirst() bool {
+	return t.first
 }
 
-// Increase increases all indices in the order by the given increment.
-func (o Order) Increase(incr int) Order {
-	for _, v := range o {
-		for i := range v {
-			v[i] += incr
+// HasNext returns true if the task has at least one next task.
+func (t *TaskNode) HasNext() bool {
+	return len(t.Next) > 0
+}
+
+// AppendToEnd adds a new task to the end of the pipeline. Note that this doesn't
+// mean that the supplied task will be executed directly after this task. Rather,
+// it means that it will be executed after all tasks in the linked list are executed.
+//
+// If any task node in the list has more than 1 next task, the function will
+// return an error, as it would be ambiguous where to append the task.
+//
+// If the task was appended successfully, it returns the created TaskNode.
+func (t *TaskNode) AppendToEnd(next ...*TaskNode) error {
+	switch len(t.Next) {
+	case 0:
+		// No next task, let's append the new task.
+		t.Next = next
+		return nil
+	case 1:
+		// Single next task, let's append the new task to it.
+		return t.Next[0].AppendToEnd(next...)
+	default:
+		// Multiple next tasks, we can't append the new task to them.
+		// If we hit this line it's an internal bug.
+		return cerrors.Errorf("(bug) multiple next tasks, please append the task to the branch where you want it")
+	}
+}
+
+// TaskNodes returns an iterator over the task nodes in the pipeline. It iterates
+// the task nodes in the order they are defined in the pipeline, depth-first.
+func (t *TaskNode) TaskNodes() iter.Seq[*TaskNode] {
+	return func(yield func(*TaskNode) bool) {
+		t.iterator()(yield)
+	}
+}
+
+// Tasks returns an iterator over the tasks in the pipeline. It iterates
+// the tasks in the order they are defined in the pipeline, depth-first.
+func (t *TaskNode) Tasks() iter.Seq[Task] {
+	return func(yield func(Task) bool) {
+		t.iterator()(func(node *TaskNode) bool {
+			return yield(node.Task)
+		})
+	}
+}
+
+// iterator is a private method that returns an iterator which also tells the
+// caller if it should stop iterating. This is needed to break the loop in parent
+// iterators, but doesn't match the Go interface for iter.Seq, so it's just a helper.
+func (t *TaskNode) iterator() func(yield func(*TaskNode) bool) bool {
+	return func(yield func(*TaskNode) bool) bool {
+		// First yield the current task.
+		if !yield(t) {
+			return false
 		}
+
+		// Then process all children in order.
+		for _, next := range t.Next {
+			if !next.iterator()(yield) {
+				return false
+			}
+		}
+		return true
 	}
-	return o
 }
 
 type ackNacker interface {

--- a/pkg/lifecycle-poc/funnel/worker.go
+++ b/pkg/lifecycle-poc/funnel/worker.go
@@ -336,7 +336,7 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 			return acker.Ack(ctx, b)
 		}
 		// There is at least one task after this one, let's continue.
-		return w.nextTask(ctx, currentIndex, b, acker)
+		return w.doNextTask(ctx, currentIndex, b, acker)
 	}
 
 	w.logger.Trace(ctx).
@@ -373,7 +373,7 @@ func (w *Worker) doTask(ctx context.Context, currentIndex int, b *Batch, acker a
 				break // break switch
 			}
 			// There is at least one task after this one, let's continue.
-			err := w.nextTask(ctx, currentIndex, subBatch, acker)
+			err := w.doNextTask(ctx, currentIndex, subBatch, acker)
 			if err != nil {
 				return err
 			}
@@ -433,7 +433,7 @@ func (w *Worker) hasNextTask(currentIndex int) bool {
 	return len(w.Order[currentIndex]) > 0
 }
 
-func (w *Worker) nextTask(ctx context.Context, currentIndex int, b *Batch, acker ackNacker) error {
+func (w *Worker) doNextTask(ctx context.Context, currentIndex int, b *Batch, acker ackNacker) error {
 	nextIndices := w.Order[currentIndex]
 	switch len(nextIndices) {
 	case 0:

--- a/pkg/lifecycle-poc/service.go
+++ b/pkg/lifecycle-poc/service.go
@@ -298,7 +298,7 @@ func (s *Service) buildRunnablePipeline(
 	pipelineLogger := s.logger
 	pipelineLogger.Logger = pipelineLogger.Logger.With().Str(log.PipelineIDField, pl.ID).Logger()
 
-	srcTasks, srcOrder, err := s.buildSourceTasks(ctx, pl, pipelineLogger)
+	srcTasks, err := s.buildSourceTasks(ctx, pl, pipelineLogger)
 	if err != nil {
 		return nil, cerrors.Errorf("failed to build source tasks: %w", err)
 	}
@@ -306,7 +306,7 @@ func (s *Service) buildRunnablePipeline(
 		return nil, cerrors.New("can't build pipeline without any source connectors")
 	}
 
-	destTasks, destOrder, err := s.buildDestinationTasks(ctx, pl, pipelineLogger)
+	destTasks, err := s.buildDestinationTasks(ctx, pl, pipelineLogger)
 	if err != nil {
 		return nil, cerrors.Errorf("failed to build destination tasks: %w", err)
 	}
@@ -314,7 +314,7 @@ func (s *Service) buildRunnablePipeline(
 		return nil, cerrors.New("can't build pipeline without any destination connectors")
 	}
 
-	procTasks, procOrder, err := s.buildProcessorTasks(ctx, pl, pl.ProcessorIDs, pipelineLogger)
+	procTasks, err := s.buildProcessorTasks(ctx, pl, pl.ProcessorIDs, pipelineLogger)
 	if err != nil {
 		return nil, cerrors.Errorf("failed to build pipeline processor tasks: %w", err)
 	}
@@ -324,18 +324,23 @@ func (s *Service) buildRunnablePipeline(
 		return nil, cerrors.Errorf("failed to build DLQ: %w", err)
 	}
 
-	tasks, order := s.combineTasksAndOrders(srcTasks, destTasks, procTasks, srcOrder, destOrder, procOrder)
+	taskNodes, err := s.buildTaskNodes(srcTasks, procTasks, destTasks)
+	if err != nil {
+		return nil, cerrors.Errorf("failed to build task nodes: %w", err)
+	}
+
+	// TODO(multi-connector): when we have multiple connectors we will have more than one task node
+	taskNode := taskNodes[0]
 
 	// log the tasks and order for debugging purposes
-	taskTypes := make([]string, len(tasks))
-	for i, task := range tasks {
-		taskTypes[i] = fmt.Sprintf("%s(%T)", task.ID(), task)
+	taskTypes := make([]string, 0)
+	for task := range taskNode.Tasks() {
+		taskTypes = append(taskTypes, fmt.Sprintf("%s(%T)", task.ID(), task))
 	}
-	pipelineLogger.Info(ctx).Any("tasks", taskTypes).Any("order", order).Msg("pipeline tasks and order")
+	pipelineLogger.Info(ctx).Any("tasks", taskTypes).Msg("pipeline tasks")
 
 	worker, err := funnel.NewWorker(
-		tasks,
-		order,
+		taskNode,
 		dlq,
 		pipelineLogger,
 		measure.PipelineExecutionDurationTimer.WithValues(pl.Config.Name),
@@ -349,46 +354,67 @@ func (s *Service) buildRunnablePipeline(
 	}, nil
 }
 
-func (s *Service) combineTasksAndOrders(
-	srcTasks, destTasks, procTasks []funnel.Task,
-	srcOrder, destOrder, procOrder funnel.Order,
-) ([]funnel.Task, funnel.Order) {
-	tasks := make([]funnel.Task, 0, len(srcTasks)+len(procTasks)+len(destTasks))
-	tasks = append(tasks, srcTasks...)
-	tasks = append(tasks, procTasks...)
-	tasks = append(tasks, destTasks...)
-
+// buildTaskNodes takes the source, processor and destination tasks and builds
+// a task node graph. The returned slice contains the first task nodes in every
+// branch of the graph. The other task nodes are connected to the first task node
+// in their branch.
+func (s *Service) buildTaskNodes(
+	srcTasks [][]funnel.Task,
+	procTasks []funnel.Task,
+	destTasks [][]funnel.Task,
+) ([]*funnel.TaskNode, error) {
 	// TODO(multi-connector): when we have multiple connectors this will not be as straight forward
-	order := srcOrder.AppendOrder(procOrder).AppendOrder(destOrder)
-	return tasks, order
+	srcTasksBranch := srcTasks[0]   // we only support one source connector for now
+	destTasksBranch := destTasks[0] // we only support one destination connector for now
+
+	taskNode := &funnel.TaskNode{Task: srcTasksBranch[0]}
+	for _, task := range srcTasksBranch[1:] {
+		err := taskNode.AppendToEnd(&funnel.TaskNode{Task: task})
+		if err != nil {
+			return nil, cerrors.Errorf("failed to append task to task node list: %w", err)
+		}
+	}
+	for _, task := range procTasks {
+		err := taskNode.AppendToEnd(&funnel.TaskNode{Task: task})
+		if err != nil {
+			return nil, cerrors.Errorf("failed to append task to task node list: %w", err)
+		}
+	}
+	for _, task := range destTasksBranch {
+		err := taskNode.AppendToEnd(&funnel.TaskNode{Task: task})
+		if err != nil {
+			return nil, cerrors.Errorf("failed to append task to task node list: %w", err)
+		}
+	}
+
+	return []*funnel.TaskNode{taskNode}, nil
 }
 
 func (s *Service) buildSourceTasks(
 	ctx context.Context,
 	pl *pipeline.Instance,
 	logger log.CtxLogger,
-) ([]funnel.Task, funnel.Order, error) {
-	var tasks []funnel.Task
-	var order funnel.Order
+) ([][]funnel.Task, error) {
+	var tasks [][]funnel.Task
 
 	for _, connID := range pl.ConnectorIDs {
 		instance, err := s.connectors.Get(ctx, connID)
 		if err != nil {
-			return nil, nil, cerrors.Errorf("could not fetch connector: %w", err)
+			return nil, cerrors.Errorf("could not fetch connector: %w", err)
 		}
 
 		if instance.Type != connector.TypeSource {
 			continue // skip any connector that's not a source
 		}
 
-		if len(tasks) > 1 {
+		if len(tasks) > 0 {
 			// TODO(multi-connector): remove check
-			return nil, nil, cerrors.New("pipelines with multiple source connectors currently not supported, please disable the experimental feature flag")
+			return nil, cerrors.New("pipelines with multiple source connectors currently not supported, please disable the experimental feature flag")
 		}
 
 		src, err := instance.Connector(ctx, s.connectorPlugins)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		srcTask := funnel.NewSourceTask(
@@ -399,48 +425,46 @@ func (s *Service) buildSourceTasks(
 		)
 
 		// Add processor tasks
-		processorTasks, processorOrder, err := s.buildProcessorTasks(ctx, pl, instance.ProcessorIDs, logger)
+		procTasks, err := s.buildProcessorTasks(ctx, pl, instance.ProcessorIDs, logger)
 		if err != nil {
-			return nil, nil, cerrors.Errorf("failed to build source processor tasks: %w", err)
+			return nil, cerrors.Errorf("failed to build source processor tasks: %w", err)
 		}
 
-		// Adjust order to include new task and the processor order
-		tasks = append(tasks, srcTask)
-		tasks = append(tasks, processorTasks...)
-
-		order = append(order, nil) // Add new task to order without attaching to previous tasks
-		order = order.AppendOrder(processorOrder)
+		// Build the slice of tasks for this source
+		srcTasks := make([]funnel.Task, 0)
+		srcTasks = append(srcTasks, srcTask)
+		srcTasks = append(srcTasks, procTasks...)
+		tasks = append(tasks, srcTasks)
 	}
 
-	return tasks, order, nil
+	return tasks, nil
 }
 
 func (s *Service) buildDestinationTasks(
 	ctx context.Context,
 	pl *pipeline.Instance,
 	logger log.CtxLogger,
-) ([]funnel.Task, funnel.Order, error) {
-	var tasks []funnel.Task
-	var order funnel.Order
+) ([][]funnel.Task, error) {
+	var tasks [][]funnel.Task
 
 	for _, connID := range pl.ConnectorIDs {
 		instance, err := s.connectors.Get(ctx, connID)
 		if err != nil {
-			return nil, nil, cerrors.Errorf("could not fetch connector: %w", err)
+			return nil, cerrors.Errorf("could not fetch connector: %w", err)
 		}
 
 		if instance.Type != connector.TypeDestination {
 			continue // skip any connector that's not a destination
 		}
 
-		if len(tasks) > 1 {
+		if len(tasks) > 0 {
 			// TODO(multi-connector): remove check
-			return nil, nil, cerrors.New("pipelines with multiple destination connectors currently not supported, please disable the experimental feature flag")
+			return nil, cerrors.New("pipelines with multiple destination connectors currently not supported, please disable the experimental feature flag")
 		}
 
 		dest, err := instance.Connector(ctx, s.connectorPlugins)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		destTask := funnel.NewDestinationTask(
@@ -451,20 +475,19 @@ func (s *Service) buildDestinationTasks(
 		)
 
 		// Add processor tasks
-		processorTasks, processorOrder, err := s.buildProcessorTasks(ctx, pl, instance.ProcessorIDs, logger)
+		procTasks, err := s.buildProcessorTasks(ctx, pl, instance.ProcessorIDs, logger)
 		if err != nil {
-			return nil, nil, cerrors.Errorf("failed to build destination processor tasks: %w", err)
+			return nil, cerrors.Errorf("failed to build destination processor tasks: %w", err)
 		}
 
-		// Adjust order to include new task and the processor order
-		tasks = append(tasks, processorTasks...)
-		tasks = append(tasks, destTask)
-
-		order = append(order, processorOrder.Increase(len(order))...) // Add processor task order without attaching to previous tasks
-		order = order.AppendSingle(nil)
+		// Build the slice of tasks for this destination
+		destTasks := make([]funnel.Task, 0)
+		destTasks = append(destTasks, destTask)
+		destTasks = append(destTasks, procTasks...)
+		tasks = append(tasks, destTasks)
 	}
 
-	return tasks, order, nil
+	return tasks, nil
 }
 
 func (s *Service) buildProcessorTasks(
@@ -472,19 +495,18 @@ func (s *Service) buildProcessorTasks(
 	pl *pipeline.Instance,
 	processorIDs []string,
 	logger log.CtxLogger,
-) ([]funnel.Task, funnel.Order, error) {
+) ([]funnel.Task, error) {
 	var tasks []funnel.Task
-	var order funnel.Order
 
 	for _, procID := range processorIDs {
 		instance, err := s.processors.Get(ctx, procID)
 		if err != nil {
-			return nil, nil, cerrors.Errorf("could not fetch processor: %w", err)
+			return nil, cerrors.Errorf("could not fetch processor: %w", err)
 		}
 
 		runnableProc, err := s.processors.MakeRunnableProcessor(ctx, instance)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		tasks = append(
@@ -496,10 +518,9 @@ func (s *Service) buildProcessorTasks(
 				s.newProcessorMetrics(pl.Config.Name, instance.Plugin),
 			),
 		)
-		order = order.AppendSingle(nil)
 	}
 
-	return tasks, order, nil
+	return tasks, nil
 }
 
 func (s *Service) buildDLQ(

--- a/pkg/lifecycle-poc/service_test.go
+++ b/pkg/lifecycle-poc/service_test.go
@@ -100,13 +100,12 @@ func TestServiceLifecycle_buildRunnablePipeline(t *testing.T) {
 		&funnel.SourceTask{},
 		&funnel.DestinationTask{},
 	}
-	is.Equal(len(got.w.Tasks), len(wantTasks))
-	for i := range got.w.Tasks {
+	i := 0
+	for got := range got.w.FirstTask.Tasks() {
 		want := wantTasks[i]
-		got := got.w.Tasks[i]
 		is.Equal(reflect.TypeOf(want), reflect.TypeOf(got)) // unexpected task type
+		i++
 	}
-	is.Equal(got.w.Order, funnel.Order{{1}, nil})
 	is.Equal(got.w.Source.(*connector.Source).Instance, source)
 }
 

